File: src/graphics/moving_balls/main_window.cpp
#include "main_window.h"
#include "movable_image.h"
#include "ui_main_window.h"

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent), ui(new Ui::MainWindow) {
  ui->setupUi(this);
  scene = new BorderedScene(QRectF(0, 0, 300, 300), 10,  this);
    ui->graphicsView->setScene(scene);

    timer = new QTimer(this);
    connect(timer, &QTimer::timeout, scene, &BorderedScene::advance);
    timer->start(100);
}

MainWindow::~MainWindow() { delete ui; }

void MainWindow::on_addButton_released() {
  scene->addItem(new MovableImage(scene->sceneRect(), this));
}

void MainWindow::on_removeButton_released() {
  for (auto item : scene->selectedItems()) {
    scene->removeItem(item);
    delete item;
  }
}

File: src/graphics/python/moving_balls/main_window.py
from PyQt6 import QtWidgets, uic, QtGui, QtCore
from PyQt6.QtCore import QRectF, QObject

from bordered_scene import BorderedScene
from movable_image import MovableImage


class MainWindow(QtWidgets.QMainWindow):
    def __init__(self):
        super().__init__()
        uic.loadUi("main_window.ui", self)
        self.scene = BorderedScene(QRectF(0, 0, 300, 300), 10, self)
        self.graphicsView.setScene(self.scene)

        timer = QtCore.QTimer(self)
        timer.timeout.connect(self.scene.advance)
        timer.start(100)

        self.show()

    def on_addButton_released(self):
        self.scene.addItem(MovableImage(self.scene.sceneRect(), self))

    def on_removeButton_released(self):
        for item in self.scene.selectedItems():
            self.scene.removeItem(item)
            del item
